### PR TITLE
Bring Clock back to build operations

### PR DESCRIPTION
**Pull request by mlopatkin**
Now Clock lives in a separate :time module, so it isn't as big of a dependency and can be used directly.

This is a prerequisite to collecting more precise timing data.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

File: platforms/core-execution/build-cache-example-client/src/main/java/org/gradle/caching/example/BuildCacheClientModule.java [hunk 1]
```diff
@@ -353,8 +354,8 @@ public void finished(BuildOperationDescriptor buildOperation, OperationFinishEve
     }
 
     @Provides
-    BuildOperationTimeSupplier createTimeSupplier() {
-        return System::currentTimeMillis;
+    Clock createTimeSupplier() {
+        return Time.clock();
```
Comment (mlopatkin) [inline platforms/core-execution/build-cache-example-client/src/main/java/org/gradle/caching/example/BuildCacheClientModule.java:358]: ❕ This signals a non-trivial change. Other clients of the library may need to be updated.

**Review by welandaz — APPROVED** [review]

File: platforms/core-runtime/build-operations/build.gradle.kts [hunk 2]
```diff
@@ -10,6 +10,7 @@ gradlebuildJava.usedInWorkers()
 dependencies {
     api(libs.jsr305)
     api(projects.stdlibJavaExtensions)
+    api(projects.time)
```
Comment (lptr) [inline platforms/core-runtime/build-operations/build.gradle.kts:13]: :x: Doing this without publishing the `:time` subproject as a library will make it impossible for the Maven cache extension to consume our libraries.

**Review by lptr — COMMENTED** [review]
Why is this necessary? How is `Clock` easier to capture than `BOTS`? I also dislike depending on something that has explicit `@ServiceScope` annotations.

**Comment by mlopatkin** [comment]
> Why is this necessary? How is `Clock` easier to capture than `BOTS`?

I'm trying to increase timestamp precision and for that I'm adding a Timestamp type in #31029. `:build-operations` has to migrate to these timestamps too, because it produces events of interest. Evolving a single `Clock` interface is tad simpler than maintaining two separate abstractions.

> I also dislike depending on something that has explicit @ServiceScope annotations.

Because scopes are tied to Gradle? What if we detach the general notion of scope from gradle-specific ones?

**Comment by lptr** [comment]
> I'm trying to increase timestamp precision and for that I'm adding a Timestamp type in https://github.com/gradle/gradle/pull/31029.

I'm all for more explicit typing, but how is wrapping a `long` containing milliseconds helpful here? How is that more precise, and what is the broader scope that requires more precision now?

Also, do we need to introduce a new type here, as opposed to using `Instant` or something? If this needs to work in workers etc., then can we wait with this change until Gradle 9 when we can rely on Java 8 available everywhere?

> Because scopes are tied to Gradle? What if we detach the general notion of scope from gradle-specific ones?

This library should be usable without Gradle, or scopes. It is a code smell that scopes appear in the library code that we should be able to remove.

**Comment by mlopatkin** [comment]
>I'm all for more explicit typing, but how is wrapping a long containing milliseconds helpful here?

Most of places just get the timestamp from the clock and pass it around as a black box. Encapsulation is an important stepping stone to avoid updating each place every time the timestamp representation changes.

> How is that more precise, and what is the broader scope that requires more precision now?

After abstracting the type we're free to change the internal representation to carry extra data.

>Also, do we need to introduce a new type here, as opposed to using Instant or something? If this needs to work in workers etc., then can we wait with this change until Gradle 9 when we can rely on Java 8 available everywhere?

- `Instant` is great, but it may not be enough. It represents a point in time with nanosecond precision, but we won't be able to add other data to it (like monotonic clock timestamps of `System.nanoTime`). 
- Besides, Gradle 9 release is at least several months away, while we may want benefits of precise tracing times sooner.

>This library should be usable without Gradle, or scopes.

Having a scope annotation doesn't prevent that. For better or worse, the annotation is necessary for Gradle's DI.

I guess we can keep the `BuildOperationTimeSupplier`, if you feel strongly about it, but convert it to produce `Timestamp` later on.

**Comment by mlopatkin** [comment]
>>How is that more precise, and what is the broader scope that requires more precision now?
>After abstracting the type we're free to change the internal representation to carry extra data.

Sorry, I've missed some extra bits here somehow. Both [trace converter](https://github.com/gradle/gradle-to-trace-converter) (that for now uses ms timestamps) and gradle profiler (that does tricks to reconstruct nanosecond precision) can benefit from these to give us more insights and integrate better with the trace viewing tools like Perfetto and Jaeger.

**Comment by lptr** [comment]
> I guess we can keep the BuildOperationTimeSupplier, if you feel strongly about it, but convert it to produce Timestamp later on.

That's not really my point. I'd rather we had fewer dependencies in build operations, is all. If we really need this, then let's do it.

**Pull request merged**